src/Microsoft.ML.Data/Transforms/KeyToValueTransform.cs

```diff
@@ -67,386 +79,513 @@ private static VersionInfo GetVersionInfo()
-
-                // REVIEW: To map the missing key to the *default* value is
-                // wrong, but the alternative is we have a bunch of null unions everywhere
-                // probably, which I am not prepared to do.
-                var defaultToken = PfaUtils.Type.DefaultTokenOrNull(TypeOutput);
-                JArray jsonValues;
-                if (TypeOutput.IsText)
+                public override JToken SavePfa(BoundPfaContext ctx, JToken srcToken)
                 {
-                    jsonValues = new JArray();
-                    for (int i = 0; i < _values.Length; ++i)
-                        jsonValues.Add(_values[i].ToString());
+                    Contracts.AssertValue(ctx);
+                    Contracts.AssertValue(srcToken);
+                    var outType = PfaUtils.Type.PfaTypeOrNullForColumnType(TypeOutput);
+                    if (outType == null)
+                        return null;
+
+                    // REVIEW: To map the missing key to the *default* value is
+                    // wrong, but the alternative is we have a bunch of null unions everywhere
+                    // probably, which I am not prepared to do.
+                    var defaultToken = PfaUtils.Type.DefaultTokenOrNull(TypeOutput);
+                    JArray jsonValues;
+                    if (TypeOutput.IsText)
+                    {
+                        jsonValues = new JArray();
+                        for (int i = 0; i < _values.Length; ++i)
+                            jsonValues.Add(_values[i].ToString());
+                    }
+                    else
+                        jsonValues = new JArray(_values);
+
+                    string cellName = ctx.DeclareCell("KeyToValueMap", PfaUtils.Type.Array(outType), jsonValues);
+                    JObject cellRef = PfaUtils.Cell(cellName);
+
+                    var srcType = Parent.InputSchema.GetColumnType(Parent.ColMapNewToOld[InfoIndex]);
+                    if (srcType.IsVector)
+                    {
+                        var funcName = ctx.GetFreeFunctionName("mapKeyToValue");
+                        ctx.Pfa.AddFunc(funcName, new JArray(PfaUtils.Param("key", PfaUtils.Type.Int)),
+                            outType, PfaUtils.If(PfaUtils.Call("<", "key", 0), defaultToken,
+                            PfaUtils.Index(cellRef, "key")));
+                        var funcRef = PfaUtils.FuncRef("u." + funcName);
+                        return PfaUtils.Call("a.map", srcToken, funcRef);
+                    }
+                    return PfaUtils.If(PfaUtils.Call("<", srcToken, 0), defaultToken, PfaUtils.Index(cellRef, srcToken));
                 }
-                else
-                    jsonValues = new JArray(_values);
+            }
+
+        }
+
+        private interface IColInput
+        {
+            PipelineColumn Input { get; }
+        }
+
+        internal sealed class OutScalarColumn<TKey, TValue> : Scalar<TValue>, IColInput
```

Comment:
>internal [](start = 8, length = 8)

If you'd just nested these inside the extensions class as we have for all other Pigsty extensions, I think you could have made these all private. At least, certainly nothing *outside* this class should be aware of these implementation.

Reply:
Makes sense

---
In reply to: [216166691](https://github.com/dotnet/machinelearning/pull/856#discussion_r216166691) [](ancestors = 216166691)